feat: Web Worker Support

wSedlacek opened the pull request:
- Doesn't assume that `window` is defined when `global` is undefined
- Exports `Override()` and `override()` so they can be imported in the cases where they aren't defined globally (like in Web Workers)

fixes: #33

- [x] Prevent Errors when running `angular-register` inside of a Web Worker
- [x] Access `@Override()` and `@override()` from the Web Worker global
- [x] ~~Register at the main entry point rather then for each worker~~
- [x] Apply the same fixes to `register`

Commit:
feat: Web Worker Support

- Doesn't assume that `window` is defined when `global` is undefined
- Exports `Override()` and `override()` so they can be imported in the cases where they aren't defined globally (like in Web Workers)

wSedlacek commented:
This is a purposed solution, that would require the import of the decorators into Web Workers.
Not entirely happy with the ergonomics but figured I would make the PR so you can take a look at it and provide some feedback.

Commit:
feat: More Universal Type Definitions

- Uses `self` rather then `window` to work inside Web Workers
- Uses `Object` rather then `any`

wSedlacek commented:
Using `self` does improve things, however since each Web Worker is still an entry point it seems that TypeScript isn't happy unless each Web Worker is registered independently.

This could just be an Angular thing though as the error is coming from the `@angular-devkit`

```
client:159 ./src/app/app.worker.ts (.../node_modules/@angular-devkit/build-angular/node_modules/worker-plugin/dist/loader.js?name=10!./src/app/app.worker.ts)
Module build failed (from .../node_modules/@angular-devkit/build-angular/node_modules/worker-plugin/dist/loader.js):
Error: apps/samples/basic/src/app/app.worker.ts:7:4 - error TS2304: Cannot find name 'Override'.
```

Commit:
feat: Expand Type Definitions to WorkerGlobalScope

hmil commented:
This is getting uncomfortably over-engineered for just the comfort of being able to write `@Override` outside of a JSDoc comment.

I can merge it if you figure out how to get it to work, but I don't think I'll have any bandwidth to look into this myself anytime soon.

Commit:
fix: Source Code Encoding

It seesm that including got screwy here at some point and resulted in a strange `Â `

This repalces that with a ` ` so errors don't occur in VS Code.

Signed-off-by: William Sedlacek <wsedlacekc@gmail.com>

Commit:
Merge branch 'encoding' into patch-1

Commit:
fix: Build time errors

TypeScript v2 doesn't seem to like this syntax
/// <reference lib="webworker" />

I would like to include this so that `WorkerGlobalScope` is properly updated

Signed-off-by: William Sedlacek <wsedlacekc@gmail.com>

wSedlacek commented:
😅  I am not super happy about it either. Give me about a week to think about it some and see if I can't come up with any other solutions.

As it stands the `self` does work inside workers, however it requires each worker to be registered.
Modifying the Global Object certainly is needlessly complicated.

wSedlacek commented:
So after doing some more research the limitation of registering with web workers globally appears to be one of Web Pack.

In this similar issue about polyfills for angular it more or less describes this https://github.com/angular/angular-cli/issues/15033
Additionally `tooling.report` seems to collaborate this issue with their failing test.
https://bundlers.tooling.report/code-splitting/

Maybe at some point in the future something like this will be supported in Web Pack and improve the ergonomics.
I do still believe that the changes made thus far in this PR are still useful as it prevents errors from occurring in the worker environment which would be required when/if Web Pack can ever do registration globally.

I will go ahead and bring the standard register up to date with this PR.

Commit:
feat: Bring `register` up to date

hmil commented on register.ts:
```diff
@@ -1,30 +1,54 @@
- */
-export function override(_target: any, _propertyKey: string, _descriptor?: PropertyDescriptor) {
-    // noop
+if (ctx) {
```
This is the core of this change, right? By checking that ctx is defined you avoid triggering an error in web workers, but at the cost of making `override` unavailable in this environment. Is that correct?

Also, if you need a solution to this, you might want to look into TypeScript transformers. They are like plugins that you can attach to the compiler to do some modifications of the AST. For instance, you can remove decorators during compilation (angular does that).

As a third alternative to JSDoc and global override symbol, a transformer could be offered to use decorators without registering a global symbol.

wSedlacek replied:
> This is the core of this change, right? By checking that ctx is defined you avoid triggering an error in web workers, but at the cost of making override unavailable in this environment. Is that correct?

The core of the change is actually the use of `self` rather then window. Self is a global property that either refers to the `Window` or the `DedicatedWorkerGlobalScope`.
```ts
typeof self === 'undefined' ? undefined : self
```

The `if` statement here is mostly a fail safe, as I have found I cannot account for every future and past environment this runs in I felt it would be best to assure that `ctx` exist prior to modifying it.

The resulting behavior is that in a `WebWorker` you can either
```ts
import 'tslint-overide/register'
// OR
import { Override } from 'tslint-overide/register'
```
Either of these should work but one of these would be required per each worker.

> Also, if you need a solution to this, you might want to look into TypeScript transformers. They are like plugins that you can attach to the compiler to do some modifications of the AST. For instance, you can remove decorators during compilation (angular does that).

I haven't worked with `transformers` before. The idea does sound interesting. This will take some time to look into.

hmil commented:
@wSedlacek Please test v1.0.1 of the package and let me know if there's any issue.

wSedlacek commented:
> @wSedlacek Please test v1.0.1 of the package and let me know if there's any issue.

Working for my use cases. Thank you!
